components/esp_http_server/src/httpd_ws.c

```diff
@@ -0,0 +1,302 @@
+// Copyright 2018 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdlib.h>
+#include <sys/random.h>
+#include <esp_log.h>
+#include <esp_err.h>
+#include <mbedtls/sha1.h>
+#include <mbedtls/base64.h>
+
+#include <esp_http_server.h>
+#include "esp_httpd_priv.h"
+
+#define TAG "httpd_ws"
+
+#define HTTPD_WS_FIN_BIT        0x80U
+#define HTTPD_WS_OPCODE_BITS    0x0fU
+#define HTTPD_WS_MASK_BIT       0x80U
+#define HTTPD_WS_LENGTH_BITS    0x7fU
```

Comment:
If these are from some standard, can you provide a reference?

Reply:
@shahpiyushv This is according to RFC6455 standard. I will add some comments here later.